### PR TITLE
[Hexagon] Resolve duplex instructions dynamically

### LINKED ISSUE #66: Replace Duplex instruction generation with sub-instruction focused approach
Duplex instructions are currently generated by taking all sub instructions and build as many permutations from them as the grouping constraints allow (see: chapter 10.3 in the v67 manual).

This approach has a lot of draw backs and no advantages (more cod needed for generation, more C code output, longer runtimes for generation and several more).

A possible replacement would be:
- Remove Duplex instruction code (in rizin and here).
- Generate struct templates for sub-instructions.
- Add a C function which is called if the parse bits are `0b00`.
- The asm string string is simply concatenated from the high and low sub instruction.
- `HexInsn` is filled with the values of the high and low sub-instruction.

Main task: `hex_disasm_with_templates` for duplex instructions is needed.

Some implementations details could be a bit tricky.
- IDs are unique for each instruction. But duplex instructions need some way to merge the IDs of both sub instructions or hold them both.

### PR DESCRIPTION
**Pull request by Rot127**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

Companion to https://github.com/rizinorg/rz-hexagon/pull/73

**Detailed description**

This PR replaces hard coded duplex instructions with sub-instructions. Duplex instructions are now resolved dynamically during disassembly.

See https://github.com/rizinorg/rz-hexagon/issues/66 for some details why this is necessary.

Main changes in this PR:
- Replaces duplex templates with sub-instruction templates.
- The `HexInsn` struct, which was passed around by all the disassembly code, is replaced by `HexInsnContainer`.
  - This also ensures a better separation between actual instruction data (like `operands`, `il_op` in `HexInsn`) and meta information (`is_duplex`, `address` in `HexInsnContainer`).
- Some refactoring like better documentation and variable renaming has been done here as well (sorry in advance. Will do it separately next time.).

This PR is also a prerequisite for the soon coming RZIL uplifting.

Note for review: many changes in the lines are simply renaming of the `hi` variable to `hic`.

**Test plan**

All current tests green.

**Closing issues**

closes https://github.com/rizinorg/rz-hexagon/issues/66

### CODE REVIEW

**Review by wargio — APPROVED** [review]
much smaller

**Comment by Rot127** [comment]
> much smaller

Don't worry. This line count comes back with with RZIL :D

File: librz/asm/arch/hexagon/hexagon.h [hunk 1]
```diff
@@ -93,33 +75,65 @@ typedef struct {
...
+	HexOp ops[HEX_MAX_OPERANDS]; ///< The operands of the instructions.
 } HexInsn;
 
+/**
+ * \brief The instruction container holds one instruction or two sub-instructions if it is a duplex.
+ * It stores meta information about those instruction(s) like opcode, packet information or the parse bits.
+ */
+typedef struct {
+	ut8 parse_bits; ///< Parse bits of instruction.
+	bool is_duplex; ///< DOes this container hold two sub-instructions?
```
Comment (XVilka) [inline librz/asm/arch/hexagon/hexagon.h]: A typo in `does`

File: librz/asm/arch/hexagon/hexagon.h [hunk 2]
```diff
@@ -93,33 +75,65 @@ typedef struct {
...
+	ut32 identifier; ///< Equals instruction ID if is_duplex = false. Otherwise: (high.id << 16) | (low.id & 0xffff)
+	union {
+		HexInsn *sub[2]; ///< Pointer to sub-instructions if is_duplex = true. sub[0] = high, sub[1] = low
+		HexInsn *insn; ///< Pointer to instruction if is_duplex = false.
+	} bin;
+	ut32 addr; ///< Address of container. Equals address of instruction or of the high sub-instruction if this is a duplex.
+	ut32 opcode; ///< The instruction opcode.
+	HexPktInfo pkt_info; ///< Packet related information. First/last instr., prefix and postfix for text etc.
+	RzAsmOp asm_op; ///< Private copy of AsmOp. Currently only of interest because it holds the utf8 flag.
```
Comment (XVilka) [inline librz/asm/arch/hexagon/hexagon.h]: Do you plan to expand its use or remove it in the future? In the latter case, I recommend adding `RZ_DEPRECATE`.

File: librz/asm/arch/hexagon/hexagon.h [hunk 3]
```diff
@@ -93,33 +75,65 @@ typedef struct {
...
 typedef struct {
-	RzList *insn; // List of instructions.
-	bool last_instr_present; // Has an instruction the parsing bits 0b11 set (is last instruction).
-	bool is_valid; // Is it a valid packet? Do we know which instruction is the first?
-	ut32 hw_loop0_addr; // Start address of hardware loop 0
-	ut32 hw_loop1_addr; // Start address of hardware loop 1
-	ut64 last_access; // Last time accessed in milliseconds
-	ut32 pkt_addr; // Address of the packet. Equals the address of the first instruction.
-	bool is_eob; // Is this packet the end of a code block? E.g. contains unconditional jmp.
+	RzList /* HexInsnContainer */ *bin; ///< Descending by address sorted list of instruction containers.
```
Comment (XVilka) [inline librz/asm/arch/hexagon/hexagon.h:130]: Here I recommend using RzVector instead, not for this PR, though.

**Review by XVilka — APPROVED** [review]

**Review by XVilka — APPROVED** [review]

File: librz/asm/arch/hexagon/hexagon.h (hunk 3 above)
Comment (Rot127) [reply librz/asm/arch/hexagon/hexagon.h:130]: https://github.com/rizinorg/rz-hexagon/issues/76

**Pull request merged**